pkg/kn/commands/service/import.go

```diff
@@ -0,0 +1,143 @@
+// limitations under the License.
+
+package service
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"knative.dev/pkg/kmeta"
+	servingv1 "knative.dev/serving/pkg/apis/serving/v1"
+
+	"github.com/spf13/cobra"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"knative.dev/client/pkg/kn/commands"
+	clientservingv1 "knative.dev/client/pkg/serving/v1"
+
+	clientv1alpha1 "knative.dev/client/pkg/apis/client/v1alpha1"
+)
+
+// NewServiceImportCommand returns a new command for importing a service.
+func NewServiceImportCommand(p *commands.KnParams) *cobra.Command {
+	var waitFlags commands.WaitFlags
+
+	command := &cobra.Command{
+		Use:   "import FILENAME",
+		Short: "Import a service and its revisions",
+		Example: `
+ # Import a service in YAML format
+ kn service import /path/to/file.yaml
+
+ # Import a service in JSON format
+ kn service import /path/to/file.json`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if len(args) != 1 {
+				return errors.New("'kn service import' requires filename of import file as single argument")
+			}
+			filename := args[0]
+
+			namespace, err := p.GetNamespace(cmd)
+			if err != nil {
+				return err
+			}
+
+			client, err := p.NewServingClient(namespace)
+			if err != nil {
+				return err
+			}
+
+			return importWithOwnerRef(client, filename, cmd.OutOrStdout(), waitFlags)
+		},
+	}
+	flags := command.Flags()
+	commands.AddNamespaceFlags(flags, false)
+	waitFlags.AddConditionWaitFlags(command, commands.WaitDefaultTimeout, "import", "service", "ready")
+
+	return command
+}
+
+func importWithOwnerRef(client clientservingv1.KnServingClient, filename string, out io.Writer, waitFlags commands.WaitFlags) error {
+	var export clientv1alpha1.Export
+	file, err := os.Open(filename)
+	if err != nil {
+		return err
+	}
+	decoder := yaml.NewYAMLOrJSONDecoder(file, 512)
+	err = decoder.Decode(&export)
+	if err != nil {
+		return err
+	}
+	if export.Spec.Service.Name == "" {
+		return fmt.Errorf("provided import doesn't content service name, please note that only kn' custom Export format is supported")
+	}
+
+	serviceName := export.Spec.Service.Name
+
+	// Return error if service already exists
+	svcExists, err := serviceExists(client, serviceName)
+	if err != nil {
+		return err
+	}
+	if svcExists {
+		return fmt.Errorf("cannot import service '%s' in namespace '%s' because the service already exists",
+			serviceName, client.Namespace())
+	}
+
+	err = client.CreateService(&export.Spec.Service)
+	if err != nil {
+		return err
+	}
+
+	// Retrieve current Configuration to be use in OwnerReference
+	retries := 0
+	var currentConf *servingv1.Configuration
+	for {
+		currentConf, err = client.GetConfiguration(serviceName)
+		if err != nil {
+			if apierrors.IsNotFound(err) && retries < 5 {
+				retries++
+				time.Sleep(time.Second)
+				continue
+			} else {
+				return err
+			}
+		}
+		break
+	}
```

Comment:
I've added a simple retry-loop to retrieve Configuration to verify it it's causing e2e test to fail. The retry can be moved to client like `UpdateServiceWithRetry`. Let me know if there are any objections to actual retry or it could be done differently.

Reply:
There's also `retry.OnError` from client-go [utils](https://github.com/kubernetes/client-go/blob/master/util/retry/util.go#L48) that's usable here, if there's a need for more sophisticated approach.

Reply:
I am with this approach; 

beside I think we can separate this retry logic in a dedicated function, but then if we'd to separate it out anyway we can consume the clint-go utils you linked. Infact we should consider using these utils wherever we find a fit to encourage consumption from these standard libs we are vendoring.  (all this can be done in a subsequent iteration if thats feasible).

Reply:
I like the idea of using standard libs. I've pushed an updated version in https://github.com/knative/client/pull/1065/commits/7b8317c366e2fb6bc1685d29ca5c6e04417dd069. We can start small here and incrementally update `UpdateServiceWithRetry` or any other instances.